qualtran/bloqs/arithmetic/hamming_weight.py

```diff
@@ -58,11 +58,13 @@ class HammingWeightCompute(GateWithRegisters):
     @cached_property
     def signature(self):
         side = Side.LEFT if self.uncompute else Side.RIGHT
+        jnk_size = self.bitsize - self.bitsize.bit_count()
+        out_size = self.bitsize.bit_length()
         return Signature(
             [
-                Register('x', self.bitsize),
-                Register('junk', self.bitsize - self.bitsize.bit_count(), side=side),
-                Register('out', self.bitsize.bit_length(), side=side),
+                Register('x', QUInt(self.bitsize)),
+                Register('junk', QBit() if jnk_size == 1 else QUInt(jnk_size), side=side),
```

Comment:
Is `QUint(1)` discouraged? This looks very verbose if we have to do this for every bloq that creates a `QUInt`

Reply:
Yeah this is sort of a decision that needs to be made (see #641). QUInt(1) and QFxp(0, ...) etc can probably be size 1. I don't think QInt(1) makes sense. #641 could move a some of the boilerplate to the constructors.

Reply:
In this particular instance, I wonder whether QAny would make more sense for a junk register, since we make no claims about what it actually is. 

For places in the library that blindly use a bitsize argument (like Add) I would prefer if we mandated that the registers are larger than 1 bit anyways; since e.g. adding a 1 bit value is a very different operation than adding n>1 bit values from the perspective of compiling and costing and algorithm for our desired architecture.

Long story short: it's a code smell if your datatypes change depending on logic like this and you should think about what your bloq is first

Reply:
> For places in the library that blindly use a bitsize argument (like Add) I would prefer if we mandated that the registers are larger than 1 bit anyways; since e.g. adding a 1 bit value is a very different operation than adding n>1 bit values from the perspective of compiling and costing and algorithm for our desired architecture. 

I think this would result in a lot of if/else blocks and code that's not generic. I agree that junk in this case should probably just be `QAny` but I don't see why it's bad to have a 1 bit integer or a 1 bit fxp number. A 1 bit fxp unsigned integer can mean a different thing compared to a 1 bit fxp number with only 1 fractional bit; and that information can be useful in the algorithm. Even if not, I think there's value in implementing classes that work for arbitrary `bitsize`.